tools/online_editor/src/editor_widget.ts

```diff
@@ -211,36 +216,46 @@ class EditorPaneWidget extends Widget {
     }
   }
 
-  public add_model(url: string, model: monaco.editor.ITextModel) {
-    if (this.#editor_documents.get(url) != null) {
-      return; // already know that URL
-    }
+  private add_model_listener(model: monaco.editor.ITextModel) {
+    console.log("New model!");
+    const uri = model.uri;
     model.onDidChangeContent(() => {
       this.maybe_update_preview_automatically();
     });
-    this.#editor_documents.set(url, { model, view_state: null });
-    this.#onModelAdded?.(url);
-    if (this.#editor_documents.size === 1) {
-      this.#base_url = url;
-      this.set_model(url);
+    this.#editor_view_states.set(uri, null);
+    this.#onModelAdded?.(uri.toString());
+    if (monaco.editor.getModels().length === 1) {
+      this.#base_url = uri.toString();
+      this.set_model(uri.toString());
       this.update_preview();
     }
   }
 
-  public remove_model(url: string) {
-    this.#editor_documents.delete(url);
-    this.#onModelRemoved?.(url);
+  public remove_model(uri: string) {
+    const uri_ = monaco.Uri.parse(uri);
+    this.#editor_view_states.delete(uri_);
+    const model = monaco.editor.getModel(uri_);
+    if (model != null) {
+      model.dispose();
+      this.#onModelRemoved?.(uri_.toString());
+    }
   }
 
-  public set_model(url: string): boolean {
-    const model_and_state = this.#editor_documents.get(url);
-    if (model_and_state != null && this.#editor != null) {
-      this.#editor.setModel(model_and_state.model);
-      if (model_and_state.view_state != null) {
-        this.#editor.restoreViewState(model_and_state.view_state);
+  public set_model(uri: string): boolean {
```

Comment:
Wouldn’t it make sense to take a Monaco uri here? Then you can avoid the string round trip in add_model_listener.